### PR TITLE
[ARCTIC-1382]Improve unit test codes in module Hive

### PR DESCRIPTION
**Pull request by zhoujinsong**
## Why are the changes needed?

As described in #1382, we need to improve the unit test codes in module Hive.

This PR attempts to replace the existing unit tests under the Hive module with a new unit testing framework, however, the test coverage of the module still falls short of the target of 90%.

Subsequent PRs will further improve the unit test coverage of the module.

## Brief change log

  - *Replace the existing unit tests under the Hive module with a new unit testing framework*
  - *Add ArcticTableUtil in module core*
  - *Make GenericTaskWriters build writer for unkeyed table*
  - *Fix BaicKeyedTableScan cannot filter expression bug*

## How was this patch tested?
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] Run test locally before making a pull request

## Documentation

  - Does this pull request introduces a new feature? (yes / **no**)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / **not documented**)

### CODE REVIEW

File: core/src/test/java/com/netease/arctic/catalog/BasicCatalogTestHelper.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+  @Override
+  public Catalog buildIcebergCatalog(CatalogMeta catalogMeta) {
+    if (!TableFormat.ICEBERG.equals(tableFormat)) {
+      throw new UnsupportedOperationException("Cannot get iceberg catalog for table format:" + tableFormat);
+    }
+    Map<String, String> catalogProperties = Maps.newHashMap(catalogMeta.getCatalogProperties());
+    catalogProperties.put(ICEBERG_CATALOG_TYPE, ICEBERG_CATALOG_TYPE_HADOOP);
+    return org.apache.iceberg.CatalogUtil.buildIcebergCatalog(TEST_CATALOG_NAME,
+        catalogProperties, new Configuration());
```
Comment (shidayang) [inline core/src/test/java/com/netease/arctic/catalog/BasicCatalogTestHelper.java:78]: This place should use ArcticCatalog, ArcticCatalog and IcebergCatalog have many behavior differences

File: hive/src/test/java/com/netease/arctic/hive/catalog/HiveTableTestHelper.java [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+  }
+
+  @Override
+  public List<DataFile> writeChangeStore(
+      KeyedTable keyedTable, long txId, ChangeAction action, List<Record> records, boolean orderedWrite) {
+    return HiveDataTestHelpers.writeChangeStore(keyedTable, txId, action, records, orderedWrite);
+  }
+
+  @Override
+  public List<DataFile> writeBaseStore(
```
Comment (shidayang) [inline hive/src/test/java/com/netease/arctic/hive/catalog/HiveTableTestHelper.java:93]: Do we need `writeHiveStore` method?

File: core/src/test/java/com/netease/arctic/catalog/BasicCatalogTestHelper.java (hunk 1 above)
Comment (zhoujinsong) [reply core/src/test/java/com/netease/arctic/catalog/BasicCatalogTestHelper.java:78]: `CatalogTestHelper#buildIcebergCatalog` is only used when performing operations on Iceberg format tables.

File: hive/src/test/java/com/netease/arctic/hive/catalog/HiveTableTestHelper.java (hunk 2 above)
Comment (zhoujinsong) [reply hive/src/test/java/com/netease/arctic/hive/catalog/HiveTableTestHelper.java:93]: `TableTestHelper` abstracted the common operations on Arctic tables.
Use `HiveDataTestHelpers#writeBaseStore` to write files in the Hive location.

File: core/src/main/java/com/netease/arctic/io/reader/AbstractArcticDataReader.java [hunk 3]
```diff
@@ -125,11 +121,15 @@ public CloseableIterator<T> readData(KeyedTableScanTask keyedTableScanTask) {
...
   public CloseableIterator<T> readDeletedData(KeyedTableScanTask keyedTableScanTask) {
-    List<PrimaryKeyedFile> equDeleteFiles = keyedTableScanTask.arcticEquityDeletes().stream()
-        .map(ArcticFileScanTask::file).collect(Collectors.toList());
-
-    if (!equDeleteFiles.isEmpty()) {
+    boolean hasDeleteFile =  keyedTableScanTask.arcticEquityDeletes().size() > 0;
+    if (!hasDeleteFile) {
+      hasDeleteFile = keyedTableScanTask.dataTasks().stream()
+          .anyMatch(arcticFileScanTask -> arcticFileScanTask.deletes().size() > 0);
+    }
```
[the remark this answers is not in the document]
Comment (zhoujinsong) [reply core/src/main/java/com/netease/arctic/io/reader/AbstractArcticDataReader.java]: SGTM. But I will keep the `hasDeleteFile` variable as the checking statement is quit long here.

File: core/src/test/java/com/netease/arctic/catalog/BasicCatalogTestHelper.java (hunk 1 above)
Comment (shidayang) [reply core/src/test/java/com/netease/arctic/catalog/BasicCatalogTestHelper.java:78]: ArcticCatalog returns ArcticTable, the FileIo of ArcticTable is ArcticFileIO, Many place use ArcticFileIO.doAs() method

File: hive/src/test/java/com/netease/arctic/hive/catalog/HiveTableTestHelper.java (hunk 2 above)
Comment (shidayang) [reply hive/src/test/java/com/netease/arctic/hive/catalog/HiveTableTestHelper.java:93]: 1

File: core/src/test/java/com/netease/arctic/catalog/BasicCatalogTestHelper.java (hunk 1 above)
Comment (zhoujinsong) [reply core/src/test/java/com/netease/arctic/catalog/BasicCatalogTestHelper.java:78]: CatalogTestHelper#buildIcebergCatalog is only used when performing operations on Iceberg format tables which `BasicIcebergCatalog` still cannot handle like creating the iceberg table.

**Review by shidayang — APPROVED** [review]

**Comment by shidayang** [comment]
LGTM

**Pull request merged**